plugins/scaffolder-backend/src/scaffolder/tasks/DefaultWorkflowRunner.ts

```diff
@@ -0,0 +1,277 @@
+      await task.getWorkspaceName(),
+    );
+    try {
+      await fs.ensureDir(workspacePath);
+      await task.emitLog(
+        `Starting up task with ${task.spec.steps.length} steps`,
+      );
+
+      const context: TemplateContext = {
+        parameters: task.spec.parameters,
+        steps: {},
+      };
+
+      for (const step of task.spec.steps) {
+        try {
+          if (step.if) {
+            const ifResult = await this.render(step.if, context);
+            if (!isTruthy(ifResult)) {
+              await task.emitLog(
+                `Skipping step ${step.id} because it's if condition was false`,
+                { stepId: step.id, status: 'skipped' },
+              );
+              continue;
+            }
+          }
+
+          await task.emitLog(`Beginning step ${step.name}`, {
+            stepId: step.id,
+            status: 'processing',
+          });
+
+          const action = this.options.actionRegistry.get(step.action);
+          const { taskLogger, streamLogger } = createStepLogger({ task, step });
+
+          const input = (step.input && this.render(step.input, context)) ?? {};
+
+          if (action.schema?.input) {
+            const validateResult = validateJsonSchema(
+              input,
+              action.schema.input,
+            );
+            if (!validateResult.valid) {
+              const errors = validateResult.errors.join(', ');
+              throw new InputError(
+                `Invalid input passed to action ${action.id}, ${errors}`,
+              );
+            }
+          }
+
+          const tmpDirs = new Array<string>();
+          const stepOutput: { [outputName: string]: JsonValue } = {};
+
+          await task.emitLog(`Beginning step ${step.name}`, {
+            stepId: step.id,
+            status: 'processing',
+          });
+
+          await action.handler({
+            baseUrl: task.spec.baseUrl,
+            input,
+            logger: taskLogger,
+            logStream: streamLogger,
+            workspacePath,
+            createTemporaryDirectory: async () => {
+              const tmpDir = await fs.mkdtemp(
+                `${workspacePath}_step-${step.id}-`,
+              );
+              tmpDirs.push(tmpDir);
+              return tmpDir;
+            },
+            output(name: string, value: JsonValue) {
+              stepOutput[name] = value;
+            },
+          });
+
+          // Remove all temporary directories that were created when executing the action
+          for (const tmpDir of tmpDirs) {
+            await fs.remove(tmpDir);
+          }
+
+          context.steps[step.id] = { output: stepOutput };
+
+          await task.emitLog(`Finished step ${step.name}`, {
+            stepId: step.id,
+            status: 'completed',
+          });
+        } catch (err) {
+          await task.emitLog(String(err.stack), {
+            stepId: step.id,
+            status: 'failed',
+          });
+          throw err;
```

Comment:
will we need to perhaps delete the tmpDirs here too? or perhaps move that to a `finally` even?